pkg/operation/botanist/etcdencryption.go

```diff
@@ -36,85 +38,197 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
-// ApplyEncryptionConfiguration creates or updates a secret on the Seed
-// which contains the encryption configuration that is necessary to encrypt the
-// Kubernetes secrets in etcd.
-//
-// To mitigate data loss to a certain degree, the secret is also synced to the Garden cluster.
-func (b *Botanist) ApplyEncryptionConfiguration(ctx context.Context) error {
-	if err := b.syncEncryptionConfigurationFromGardenCluster(ctx); err != nil {
+// SyncEncryptionConfiguration syncs an already generated etcd encryptio	n configuration secret to the shootstate
+// TODO: This step can be removed in the future, once all shoots have started using the shoot state to persist resources.
+func (b *Botanist) SyncEncryptionConfiguration(ctx context.Context) error {
+	secret := &corev1.Secret{ObjectMeta: kutil.ObjectMeta(b.Shoot.SeedNamespace, common.EtcdEncryptionSecretName)}
+	err := b.K8sSeedClient.Client().Get(ctx, kutil.KeyFromObject(secret), secret)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
 		return err
 	}
-	secret, err := b.createOrUpdateEncryptionConfiguration(ctx)
+	shootState := &gardencorev1alpha1.ShootState{ObjectMeta: kutil.ObjectMeta(b.Shoot.Info.Namespace, b.Shoot.Info.Name)}
+	_, err = controllerutil.CreateOrUpdate(ctx, b.K8sGardenClient.Client(), shootState, func() error {
+		gardenerResourceDataList := gardencorev1alpha1helper.GardenerResourceDataList(shootState.Spec.Gardener)
+		encryptionKey, err := infodata.GetInfoData(gardenerResourceDataList, common.ETCDSecretsEncryptionKeyDataName)
+		if err != nil {
+			return err
+		}
+		secretsEncryptionState, err := infodata.GetInfoData(gardenerResourceDataList, common.ETCDSecretsEncryptionStateName)
+		if err != nil {
+			return err
+		}
+
+		if encryptionKey == nil {
+			conf, err := encryptionconfiguration.ReadSecret(secret)
+			name, key, err := encryptionconfiguration.GetSecretKeyForResources(conf, common.EtcdEncryptionEncryptedResourceSecrets)
+			if err != nil {
+				return err
+			}
+			encryptionKey = &encryptionconfiguration.ETCDEncryptionKey{
+				Key:  key,
+				Name: name,
+			}
+			infodata.UpsertInfoData(&gardenerResourceDataList, common.ETCDSecretsEncryptionKeyDataName, encryptionKey)
+		} else {
```

Comment:
This else branch can never be reached, right?

Reply:
On second reconciliations they will no longer be nil, but come to think of it I don't see it as completely necessary to log this on every reconciliation.